sdk/cxx/include/sawtooth/transaction_handler.h

```diff
@@ -0,0 +1,118 @@
+/* Copyright 2017 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <map>
+#include <memory>
+#include <list>
+#include <string>
+
+#include "proto/transaction.pb.h"
+
+#include "sawtooth/global_state.h"
+
+namespace sawtooth {
+
+// Utility types to allow for managed instances of std::string on the heap.
+typedef std::unique_ptr<std::string> StringUPtr;
+typedef std::shared_ptr<std::string> StringPtr;
+
+// The transaction data for a Transaction Processing request.
+class Transaction final {
+ public:
+    Transaction(StringPtr header, StringPtr payload, StringPtr signature):
+            payload_(payload), signature_(signature) {
+        this->header_.ParseFromArray(header->c_str(),
+            header->length());
+    }
+    Transaction (const Transaction&) = delete;
+    Transaction (const Transaction&&) = delete;
+    Transaction& operator= (const Transaction&) = delete;
+
+    const TransactionHeader& header() const {
+        return this->header_;
+    }
+
+    const std::string& payload() const {
+        return *(this->payload_);
+    }
+
+    const std::string& signature() const {
+        return *(this->signature_);
+    }
+
+ private:
+    TransactionHeader header_;
+    StringPtr payload_;
+    StringPtr signature_;
+};
+typedef std::unique_ptr<Transaction> TransactionUPtr;
+
+
+// TransactionApplicator is an interface that defined to represent the
+// processing of a single transaction. This object should make no assumptions
+// about the thread and timing of when it will execute. The
+// TransactionApplicator owns the reference to the transaction and the global
+// state it is passed.
+class TransactionApplicator {
+ public:
+    TransactionApplicator(sawtooth::TransactionUPtr txn,
+        sawtooth::GlobalStateUPtr state) : txn(std::move(txn)),
+        state(std::move(state)) {}
+    ~TransactionApplicator() {}
```

Comment:
This needs to be marked `virtual` since you have a pure virtual function and the intent is that the class should be derived from.  Deleting from the base class pointer will lead to shearing.